clash_lib/src/proxy/tproxy/mod.rs

```diff
@@ -15,26 +17,202 @@ impl Drop for Listener {
+                ..Default::default()
+            };
+
+            trace!("tproxy new tcp conn {}", sess);
+
+            let dispatcher = self.dispather.clone();
+            tokio::spawn(async move {
+                dispatcher.dispatch_stream(sess, socket).await;
+            });
+        }
     }
 
+    #[cfg(not(target_os = "linux"))]
     async fn listen_tcp(&self) -> std::io::Result<()> {
-        unimplemented!("don't listen to me :)")
+        warn!("tproxy not supported on non Linux");
+        Ok(())
+    }
+
+    #[cfg(target_os = "linux")]
+    async fn listen_udp(&self) -> std::io::Result<()> {
+        use std::os::fd::{AsFd, AsRawFd};
+
+        use socket2::{Domain, Socket};
+
+        let socket = Socket::new(Domain::IPV4, socket2::Type::DGRAM, None)?;
+        socket.set_ip_transparent(true)?;
+        socket.set_nonblocking(true)?;
+        socket.set_broadcast(true)?;
+
+        let enable = 1u32;
+        let payload = std::ptr::addr_of!(enable).cast();
+        unsafe {
+            libc::setsockopt(
+                socket.as_fd().as_raw_fd(),
+                libc::IPPROTO_IP,
+                libc::IP_RECVORIGDSTADDR,
+                payload,
+                std::mem::size_of_val(&enable) as libc::socklen_t,
+            )
+        };
+        socket.bind(&self.addr.into())?;
+
+        let listener = unix_udp_sock::UdpSocket::from_std(socket.into())?;
+
+        handle_inbound_datagram(Arc::new(listener), self.dispather.clone()).await
     }
 
+    #[cfg(not(target_os = "linux"))]
     async fn listen_udp(&self) -> std::io::Result<()> {
-        unimplemented!("don't listen to me :)")
+        warn!("tproxy not supported on non Linux");
+        Ok(())
     }
 }
+
+#[cfg(target_os = "linux")]
+async fn handle_inbound_datagram(
+    socket: Arc<unix_udp_sock::UdpSocket>,
+    dispatcher: Arc<Dispatcher>,
+) -> std::io::Result<()> {
+    use tracing::trace;
+
+    use crate::{
+        proxy::datagram::UdpPacket,
+        session::{Network, Session, Type},
+    };
+
+    use super::tun::TunDatagram;
+
+    // dispatcher <-> tproxy communications
+    let (l_tx, mut l_rx) = tokio::sync::mpsc::channel(32);
+
+    // forward packets from tproxy to dispatcher
+    let (d_tx, d_rx) = tokio::sync::mpsc::channel(32);
+
+    // for dispatcher - the dispatcher would receive packets from this channel,
+    // which is from the stack and send back packets to this channel, which is
+    // to the tproxy
+    let udp_stream = TunDatagram::new(l_tx, d_rx);
+
+    let sess = Session {
+        network: Network::Udp,
+        typ: Type::Tproxy,
+        ..Default::default()
+    };
+
+    let closer = dispatcher.dispatch_datagram(sess, Box::new(udp_stream));
+
+    // dispatcher -> tproxy
+    let responder = socket.clone();
+    let fut1 = tokio::spawn(async move {
+        while let Some(pkt) = l_rx.recv().await {
+            trace!("tproxy <- dispatcher: {:?}", pkt);
+
+            // remote -> local
+            match responder
```

Comment:
@VendettaReborn we don't need to create a new response_transparent socket here and can just use the original listen socket to response